gms/gossiper.cc

```diff
@@ -1133,7 +1133,7 @@ future<> gossiper::assassinate_endpoint(sstring address) {
                 auto ring_delay = std::chrono::milliseconds(gossiper._cfg.ring_delay_ms());
                 logger.info("Sleeping for {} ms to ensure {} does not change", ring_delay.count(), endpoint);
                 // make sure it did not change
-                sleep(ring_delay).get();
+                sleep_abortable(ring_delay, gossiper._abort_source).get();
```

Comment:
This sleep is part of the `assassinate_endpoint()` operation (the operation joins with this, it doesn't run in the background). The sleeps in `wait_for_gossip_to_settle()` are also like that. In the latter case we chose the caller to interrupt the operation (by providing an abort_source as an argument). In the former case, we chose stop_gossiping() to interrupt it. We could have let the caller interrupt `assasinate_endpoint()` as well. I wonder why did you chose otherwise.

The idea was that whoever uses the gossiper (having an active foreground operation counts as a use), should be stopped before the gossiper is stopped. And when the gossiper is stopped, it interrupts only operations started by itself, which don't count as external use.

Reply:
I converted other methods to take an abort source as well. The only internal task I see is `gossiper::run` and it does not sleep, so we may drop the `abort_source` altogether.

I observed that we do not wait for HTTP requests to finish before stopping the gossiper, so it is possible for somebody to start, for example, an `assasinate` operation while we are shutting down. I'm not sure what to do about it - guard such operations with a gate?

Reply:
Before the gossiper is stopped (sharded::stop), the API server should be stopped. Or at least the part of the API which uses the gossiper. Stopping is interrupt (abort_source) + await (gate for tracking).

Reply:
Proper shutdown of services seems to be a large topic (https://github.com/scylladb/scylla/issues/293). I think this is a topic too big to be handled here. I tried calling `sharded::stop` on the gossip service, but some other services still tried to use it during shutdown, which caused the nodes on my local cluster to crash.

I don't see a way to unregister a subset of HTTP routes, so I added a gate for gossiper methods that face the HTTP API. The gate is closed during shutdown operation (via `defer_verbose_shutdown`). Those routes will just return an error in case they are called during shutdown.